File: python/GC/burn_tester.py
import const
import threading
import const
import garbage_allocator
import set_allocator
import random
import baseline_test
import sampler
import datetime 
import common.distribution_transforms as cd
import gc
import time
import interval 
import math

results = [None]*const.Threadcount
NanosecondsPerReleaseCycle = 1000
TimeSamplerFrequency = 1000000
AllocationSequenceLength = 2**20

NanoToMilliseconds = 1000/ 1e9
NanoToseconds = 1 / 1e9

# Result container
class GcResult:
    def __init__(self,
    Duration,
    staticSetSize,
    AllocationSpeed,
    ThreadCount,
    MaxSize,
    MaxTime,
    OperationPerSecond,
    RAMUseBefore,
    RAMUseAfter,
    GCrate,
    globalPauses):
    self.Duration = Duration
    self.staticSetSize = staticSetSize
    self.AllocationSpeed = AllocationSpeed
    self.ThreadCount = ThreadCount
    self.MaxSize = MaxSize
    self.MaxTime = MaxTime
    self.OperationPerSecond = OperationPerSecond
    self.RAMUseBefore = RAMUseBefore
    self.RAMUseAfter = RAMUseAfter
    self.GCrate = GCrate
    self.globalPauses = globalPauses

class BurnTester:
    def __init__(self,
        # flag to signal whether to give output or not
        NoOutput,
        #
        Duration,
        MaxTime,
        MaxSize,
        StaticSetSize,
        Allocations,
        StartIndexes,
        StaticSet,
        StaticSetCount,
        Random,
        isInitialized 
         ):
        self.NoOutput = NoOutput
        self.Duration = Duration
        self.MaxTime = MaxTime
        self.MaxSize = MaxSize
        self.StaticSetSize = StaticSetSize
        self.Allocations = Allocations
        self.StartIndexes = StartIndexes
        self.StaticSet = StaticSet
        self.StaticSetCount = StaticSetCount
        self.Random = Random
        self.isInitialized = isInitialized


# In this thread we create GarbageAllocator
def _AllocatorThread(tester,index, startOffset):
     global results
     results[index] = set_allocator.NewSetAllocator((tester.StaticSetSIze)/const.Threadcount, 
     tester.Allocations, 
     (startOffset + tester.StartIndexes[index])%const.AllocationSequenceLength)
    
def _GarbageAllocatorThread(tester, index, ):
    results[index] = garbage_allocator.NewGabageAllocator(tester.Duration, tester.Allocations, tester.StartIndexes[index])

def NewGCTester (staticSetSize):
    return BurnTester(False, 
        const.DefaultDuration, 
        const.DefaultMaxTime, 
        const.DefaultMaxSize,
        staticSetSize,
        [garbage_allocator.AllocationInfo]*AllocationSequenceLength,
        # This actually should be int32 but i guess it is fine to do this for now since it is python..
        []*AllocationSequenceLength,
        [[[]]],
        0,
        random.NewStdRandom(123),
        False )

def NewWarmupGCTester (staticSetSize):
    t = NewGCTester(staticSetSize)
    t.NoOutput = True
    t.Duration = datetime.timedelta(second =1)


def TryInitialize(BurnTester):
    if BurnTester.isinitialized:
        return
    sizeSampler = sampler.CreateStandardSizeSampler(BurnTester.Random)
    
    timeSampler = sampler.CreateStandardTimeSampler(BurnTester.Random)
    releaseCycleTimeInSeconds = 1e-6
    timeFactor = 1 / TimeSamplerFrequency / releaseCycleTimeInSeconds
    for i in range(AllocationSequenceLength):
        size = cd.truncate(cd.truncateMin((sizeSampler.Sample(),0)),1,BurnTester.MaxSize)
        arraySize = ByteSizeToArraySize(size)
        _time = cd.truncate(cd.truncateMin((timeSampler.Sample(),0)),0,BurnTester.MaxSize)
        _timeStr = str(int(_time*timeFactor))
        bucketIndex = int(len(_timeStr) -1 )
        generationIndex = int(_timeStr[0]-'0')

        BurnTester.Allocations[i] = garbage_allocator.AllocationInfo(arraySize, bucketIndex, generationIndex)

    # Remember to deal with the thread information here
    for i in range(const.Threadcount):
        BurnTester.StartIndexes[i] = random.Next(BurnTester.Random) % AllocationSequenceLength
    
    startOffset = random.Next(BurnTester.Random) % AllocationSequenceLength

    
    # start the multithreading here
    threads = [
            threading.Thread(target=_AllocatorThread, args=(BurnTester, i, startOffset ),daemon=True)
            for i in range(const.Threadcount)
            ]
    for t in threads:
        t.start()

    for a in results:
        allocator = a.SetAllocator
        BurnTester.StaticSet.append(allocator.Set)
        BurnTester.StaticSetCount = BurnTester.StaticSetCount + len(allocator.Set)
    
    # Force to do a gc here
    gc.collect()
    BurnTester.isInitialized = True

def run_gc_tester(BurnTester):
    TryInitialize(BurnTester)
    
    # Normalize GCPause 
    allocators = []

    #
    for a in results:
        allocators.append(a.GarbageAllocator)
    
    for a in range(allocators):
        d = a.EndTimestamp- startTime
        if duration < d:
            duration = d
        
        for j in range(a.GCPauses):
            
    #
    threads = [
        threading.Thread(target=_GarbageAllocatorThread, args=(tester, i),daemon=True)
        for i in range(const.Treadcount)
        ]
    startTime = time.time()
    for t in threads:
        t.start()
    
    pauses = []
    duration = 0
    for a in results:

        d = (a.EndTimestamp-startTime) * NanoToseconds
        if duration < d:
            duration = d
        
        for j in range(a.GCPauses):
            a.GCPauses[j].Start = (a.GCPauses[j].Start - startTime) * NanoToMilliseconds
            a.GCPauses[j].End = (a.GCPauses[j].End - startTime - 0.5) * NanoToMilliseconds

        a.GCPauses = interval.ToCanonicalSorted(a.GCPauses)
        pauses.append(a.GCPauses)
    
    intersections = pauses[0]
    for p in pauses:
        intersections = interval.IntersectSortedPairs(intersections, p)
    
    globalPauses = []
    globalPausesSum = 0
    for p in intersections:
        globalPauses.append(p.End - p.Start)
        globalPausesSum = globalPausesSum + (p.End - p.Start)

    allocationSizes = []
    allocationHoldDuration = []
    for a in allocators:
        for ai in a.allocations:
            t = math.Pow(10, ai.BucketIndex)*ai.GenerationIndex * NanosecondsPerReleaseCycle * NanoToMilliseconds
            allocationHoldDuration.append(t)
            allocationSizes.append(garbage_allocator.ArraySizeToByteSize(ai.ArraySize))
    
    ops = 0
    bytes = 0
    for a in allocators:
        ops = a.AllocationCount + ops
        bytes = a.AllocationCOunt + ops
    # TODO: I cannot find an efficient way to find out the 

File: Go/burn_test/burn_tester.go
package burn_test

import (
	. "../common"
	. "../common/nanotime"
	"fmt"
	"math"
	"runtime"
	"strconv"
	"time"
)

const ArrayItemSize = 8
const ArrayItemSizeLog2 = 3

var MinArraySize = int32(24)
var MinAllocationSize = int32(MinArraySize + 8)

const TimeSamplerFrequency = 1000000
const AllocationSequenceLength = 1 << 20 // 2^20, i.e. ~ 1M items; must be a power of 2!

var MinGCPause = time.Duration(10 * time.Microsecond).Nanoseconds()
var DefaultDuration = time.Duration(10 * time.Second)
var DefaultMaxTime = float64(1000) * TimeSamplerFrequency // 1000 seconds
var DefaultMaxSize = float64(1 << 17)                     // 128KB
var releaseCycleTimeInSeconds = NanosecondsPerReleaseCycle * Nano
var TimeFactorFactor = float64(1)
var TimeFactor = ((1.0 / TimeSamplerFrequency) / releaseCycleTimeInSeconds) * TimeFactorFactor

type GcResult struct {
	//in seconds
	Duration float64
	// static set size, GB
	staticSetSize float64
	// Allocation Speed, operation per second
	AllocationSpeed float64
	// Thread Count
	ThreadCount int
	//Maxsize
	MaxSize int32
	MaxTime float64
	// Operationpersecond
	OperationperSecond float64
	// RAM used
	RAMUseBefore float64
	RAMUseAfter  float64
	GCrate       float64
	globalPauses float64
}
type BurnTester struct {
	NoOutput       bool
	Duration       time.Duration
	MaxTime        float64
	MaxSize        float64
	StaticSetSize  int64
	Allocations    []AllocationInfo
	StartIndexes   []int32
	StaticSet      [][][]int64
	StaticSetCount int64
	Random         *StdRandom
	isInitialized  bool
}

func NewBurnTester(staticSetSize int64) *BurnTester {
	t := &BurnTester{
		NoOutput:       false,
		Duration:       DefaultDuration,
		MaxTime:        DefaultMaxTime,
		MaxSize:        DefaultMaxSize,
		StaticSetSize:  staticSetSize,
		Allocations:    make([]AllocationInfo, AllocationSequenceLength),
		StartIndexes:   make([]int32, AllocationSequenceLength),
		StaticSet:      make([][][]int64, 0),
		StaticSetCount: 0,
		Random:         NewStdRandom(123),
	}
	return t
}

func NewWarmupBurnTester(staticSetSize int64) *BurnTester {
	t := NewBurnTester(staticSetSize)
	t.NoOutput = true
	t.Duration = time.Duration(1 * time.Second)
	return t
}

func (t *BurnTester) TryInitialize() float64 {
	if t.isInitialized {
		return 0
	}
	// sizeSampler is a transformeddistribution
	//transformedDistribution.sample ->

	// Truncate ->
	// TransformedDistribution(
	// createstandardsizesampler_result,
	// func(x float64) float64 {return math.Max(min, math.Min(max, x))}
	//)

	// I guess this will work as a recursion

	// Therefore it will first do max()

	// The return value of CreateStandardSizeSampler
	// TruncateMin(combinedDistribution,
	// func(x float64) float64 {return math.Max(x. 0)})
	sizeSampler := Truncate(CreateStandardSizeSampler(t.Random), 1, t.MaxSize)
	timeSampler := Truncate(CreateStandardTimeSampler(t.Random), 0, t.MaxTime)

	for i := 0; i < AllocationSequenceLength; i++ {
		// math.Max(min, math.Min(max, createstandardsizesampler.sample()))
		size := int32(sizeSampler.Sample())

		arraySize := ByteSizeToArraySize(size)
		// TransformedDistribution.sample -> t.transform(t.d.sample())
		_time := timeSampler.Sample()
		//Itoa int to string
		_timeStr := strconv.Itoa(int(_time * TimeFactor))
		bucketIndex := int8(len(_timeStr) - 1)
		generationIndex := int8(_timeStr[0] - '0')

		t.Allocations[i] = AllocationInfo{arraySize, bucketIndex, generationIndex}
	}

	for i := 0; i < ThreadCount; i++ {
		t.StartIndexes[i] = int32(t.Random.Next()) % AllocationSequenceLength
	}

	startOffset := int32(t.Random.Next()) % AllocationSequenceLength
	runner := NewParallelRunner(func(i int) IActivity {
		return NewSetAllocator(
			t.StaticSetSize/int64(ThreadCount),
			t.Allocations,
			(startOffset+t.StartIndexes[i])%AllocationSequenceLength)
	})

	// Set allocator here is used to allocate static set

	activities := runner.Run()
	for _, a := range activities {
		allocator := a.(*SetAllocator)
		t.StaticSet = append(t.StaticSet, allocator.Set)
		t.StaticSetCount += int64(len(allocator.Set))
	}

	runtime.GC()

	t.isInitialized = true
	return TimeFactor
}

func (t *BurnTester) Run() *GcResult {
	var timefactor = t.TryInitialize()
	// In this try initialize process, allocation[] is filled with the randomization result

	// This variable here does not work
	testDuration := t.Duration.Seconds()
	if !t.NoOutput {
		fmt.Printf("Test settings:\n")
		fmt.Printf("  Duration:          %v s\n", int(testDuration))
		fmt.Printf("  Thread count:      %v\n", ThreadCount)
		fmt.Printf("  Static set:\n")
		fmt.Printf("    Total size:      %.3f GB\n", float64(t.StaticSetSize)/GB)
		fmt.Printf("    Object count:    %.3f M\n", float64(t.StaticSetCount)/Mega)
		fmt.Println()
	}

	runner := NewParallelRunner(func(i int) IActivity { return NewGarbageAllocator(t.Duration, t.Allocations, t.StartIndexes[i]) })

	msPre := new(runtime.MemStats)
	runtime.ReadMemStats(msPre)
	startTime := float64(Nanotime().Nanoseconds())
	activities := runner.Run()
	msPost := new(runtime.MemStats)
	runtime.ReadMemStats(msPost)

	// Slice item casting
	allocators := make([]*GarbageAllocator, 0)
	for _, a := range activities {
		allocators = append(allocators, a.(*GarbageAllocator))
	}

	const NanoToMilliseconds = 1000 / Giga // ns to ms
	const NanoToSeconds = 1 / Giga         // ns to ms

	// Normalizing GCPauses; nanoseconds to milliseconds and a bit more
	var pauses [][]Interval
	var duration float64
	for _, a := range allocators {
		d := (float64(a.EndTimestamp) - startTime) * NanoToSeconds
		if duration < d {
			duration = d
		}
		for j := range a.GCPauses {
			a.GCPauses[j].Start = (a.GCPauses[j].Start - startTime) * NanoToMilliseconds
			a.GCPauses[j].End = (a.GCPauses[j].End - startTime - 0.5) * NanoToMilliseconds
		}
		a.GCPauses = ToCanonicalSorted(a.GCPauses)
		pauses = append(pauses, a.GCPauses)
	}
	var intersections = pauses[0]
	for _, p := range pauses {
		intersections = IntersectSortedPairs(intersections, p)
	}
	var globalPauses []float64
	var globalPausesSum float64
	for _, p := range intersections {
		globalPauses = append(globalPauses, p.End-p.Start)
		globalPausesSum += p.End - p.Start
	}
	var allocationSizes []float64
	var allocationHoldDurations []float64
	for _, a := range allocators {
		for _, ai := range a.Allocations {
			t := math.Pow(10, float64(ai.BucketIndex)) * float64(ai.GenerationIndex) * NanosecondsPerReleaseCycle * NanoToMilliseconds
			allocationHoldDurations = append(allocationHoldDurations, t)
			allocationSizes = append(allocationSizes, float64(ArraySizeToByteSize(ai.ArraySize)))
		}
	}

	fmt.Printf("Actual duration:         %.3f s\n", duration)
	fmt.Printf("Allocation speed:\n")
	var ops, bytes int64
	for _, a := range allocators {
		ops += a.AllocationCount
		bytes += a.ByteCount
	}
	fmt.Printf("  Operations per second: %.2f M/s\n", float64(ops)/duration/Mega)
	fmt.Printf("  Bytes per second:      %.2f GB/s\n", float64(bytes)/duration/GB)
	fmt.Printf("  Allocation stats:\n")
	fmt.Printf("    Size:\n")
	DumpArrayStats(allocationSizes, "B", "      ", true)
	fmt.Printf("    Hold duration:\n")
	DumpArrayStats(allocationHoldDurations, "ms", "      ", true)
	fmt.Println()
	fmt.Printf("GC stats:\n")
	fmt.Printf("  RAM used:              %.3f -> %.3f GB\n", float64(msPre.HeapAlloc)/GB, float64(msPost.HeapAlloc)/GB)
	fmt.Printf("  GC rate:               %.3f /s\n", float64(msPost.NumGC-msPre.NumGC)/duration)
	fmt.Printf("  Allocation rate:\n")
	fmt.Printf("    Objects:             %.3f M/s, %.3f M/s freed\n",
		float64(msPost.Mallocs-msPre.Mallocs)/duration/Mega,
		float64(msPost.Frees-msPre.Frees)/duration/Mega)
	fmt.Printf("    Bytes:               %.3f GB/s\n", float64(msPost.TotalAlloc-msPre.TotalAlloc)/duration/GB)
	fmt.Printf("  Global pauses:\n")
	fmt.Printf("    %% of time frozen:    %.3f %%, %.3f %% as reported by runtime\n",
		globalPausesSum/1000/duration*100,
		float64(msPost.PauseTotalNs-msPre.PauseTotalNs)/Giga/duration*100)
	fmt.Printf("    # per second:        %.3f /s\n", float64(len(globalPauses))/duration)
	fmt.Printf("    Pause duration:\n")
	DumpArrayStats(globalPauses, "ms", "      ", true)
	fmt.Println()

	var staticsetsize = t.StaticSetSize
	return &GcResult{
		Duration:           duration,
		staticSetSize:      float64(staticsetsize) / GB,
		AllocationSpeed:    float64(msPost.Mallocs-msPre.Mallocs) / duration / Mega,
		ThreadCount:        ThreadCount,
		MaxSize:            MinArraySize,
		MaxTime:            timefactor,
		OperationperSecond: float64(ops) / duration / Mega,
		RAMUseBefore:       float64(msPre.HeapAlloc) / GB,
		RAMUseAfter:        float64(msPost.HeapAlloc) / GB,
		GCrate:             float64(msPost.NumGC-msPre.NumGC) / duration,
		globalPauses:       globalPausesSum / 1000 / duration * 100,
	}
}
